### PR TITLE
Fix applying fog to ssao

### PR DESCRIPTION
**Pull request by Hadegda**
## Description

- fix appearing molecule footprints when ssao is on (they was visible then fog was disabled)
- rewrite formula of fog applying to ssao (now it's logically correct: pixColor*AO*(1-fogFactor) + fogColor*fogFactor)
- fix code issues

## Type of changes
- Bug fix (non-breaking change which fixes an issue)

## Checklist
- [x] I have read [CONTRIBUTING](https://github.com/epam/miew/blob/master/CONTRIBUTING.md) and [CODE_OF_CONDUCT](https://github.com/epam/miew/blob/master/CODE_OF_CONDUCT.md) guides.
- [x] I have followed the code style of this project.
- [x] I have run `yarn run ci`: lint and tests pass locally with my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added necessary documentation / The changes do not need docs update.

### CODE REVIEW

File: src/gfx/gfxutils.js [hunk 1]
```diff
@@ -196,6 +196,25 @@ THREE.Raycaster.prototype.intersectVisibleObject = function (gfxObj, camera, cli
   return p;
 };
 
+THREE.Matrix4.prototype.extractScale = (function () {
+  const _v = new THREE.Vector3();
+
+  return function (scale) {
+    if (scale === undefined) scale = _v.clone();
```
Comment (NataliaDSmirnova) [inline src/gfx/gfxutils.js]: I think we should follow one of two ways here:
1. don't create new vector if `scale` is `undefined `and simply return 
2. create new vector for scale but at the same time log a warning with the fact, that `new` is too expensive operation to do it on-the-fly.
I would prefer the first one, what about you?

File: src/gfx/gfxutils.js [hunk 2]
```diff
@@ -196,6 +196,25 @@ THREE.Raycaster.prototype.intersectVisibleObject = function (gfxObj, camera, cli
   return p;
 };
 
+THREE.Matrix4.prototype.extractScale = (function () {
```
Comment (NataliaDSmirnova) [inline src/gfx/gfxutils.js]: It's great to have such a function. Is there any other place in code to use it?

File: src/gfx/shaders/AO.frag [hunk 3]
```diff
@@ -49,6 +48,13 @@ void main() {
   vec3 viewPos = ViewPosFromDepth(vUv);
   // remap coordinates to prevent noise exture rescale
   vec2 vUvNoise = vUv / srcTexelSize * noiseTexelSize;
+  vec4 normalData = texture2D(normalTexture, vUv);
+  // return for background fragments (their normals are zero vectors)
+  if (length(normalData.rgb) < 0.000000001) {
```
Comment (NataliaDSmirnova) [inline src/gfx/shaders/AO.frag]: As I know  we use float type here, why do you compare with the 9th digit after a point?

File: src/gfx/shaders/AOHorBlur.frag [hunk 4]
```diff
@@ -12,14 +12,25 @@ void main() {
   float x = vUv.x;
   float y = vUv.y;
   vec4 res = vec4(0.0);
+  res.a = texture2D(aoMap, vec2(x, y )).a;
+  // return for background fragments (0.0 in alpha component means that it is background fragment)
+  if (res.a < 0.000000001) {
```
Comment (NataliaDSmirnova) [inline src/gfx/shaders/AOHorBlur.frag]: I ask the same question here :)

File: src/gfx/shaders/AOHorBlur.frag [hunk 5]
```diff
@@ -12,14 +12,25 @@ void main() {
...
+    gl_FragColor = res;
+    return;
+  }
+
   float pixelDepth = texture2D(depthTexture, vec2(x, y)).x;
   float weightSum = 0.0;
   for (int i = 0; i < MAX_SAMPLES_COUNT; ++i) {
+    if (texture2D(aoMap, vec2(x + samplesOffsets[i] * srcTexelSize.x, y )).a < 0.000000001) {
```
Comment (NataliaDSmirnova) [inline src/gfx/shaders/AOHorBlur.frag]: and here :)

File: src/gfx/shaders/AOMaterial.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    // set default values
+    this.setValues.call(this, {
+      uniforms: {
+        noiseTexture: { type: 't', value: noise.noiseTexture },
+        noiseTexelSize: { type: 'v2', value: new THREE.Vector2(1.0 / noise.noiseWidth, 1.0 / noise.noiseHeight) },
+        diffuseTexture: { type: 't', value: null },
+        normalTexture: { type: 't', value: null },
+        depthTexture: { type: 't', value: null },
+        srcTexelSize: { type: 'v2', value: new THREE.Vector2(1.0 / 512.0, 1.0 / 512.0) },
+        camNearFar: { type: 'v2', value: new THREE.Vector2(1.0, 10.0) },
```
Comment (NataliaDSmirnova) [inline src/gfx/shaders/AOMaterial.js]: Do we still use the uniform `camNearFar`. I've found its declaration in AO.frag, but it isn't used. Let's remove it!

File: src/gfx/shaders/AOVertBlurWithBlend.frag [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    // add fog to the result value
+    // goal:    gl_FragColor = fragColor*AO*(1-fogFactor) + fogColor*fogFactor
+    // exists:  AO, fogFactor, fogColor,
+    //          color = fragColor*(1-fogFactor) + fogColor*fogFactor
+    // simplify:
+    //          fragColor*AO*(1-fogFactor) + fogColor*fogFactor =
+    //        = [fragColor*(1-fogFactor) = color - fogColor*fogFactor] =
+    //        = (color - fogColor*fogFactor)*AO + fogColor*fogFactor =
+    //        = color*AO + fogColor*fogFactor*(1 - AO)
+    // result:  gl_FragColor = color*AO + fogColor*fogFactor*(1 - AO)
```
Comment (NataliaDSmirnova) [inline src/gfx/shaders/AOVertBlurWithBlend.frag]: I think we should add here information about color, for example, `color `is gotten from `diffuseTexture`, where molecule has been already drawn with fog.

File: src/gfx/shaders/AOVertBlurWithBlend.frag [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    float depth = texture2D(depthTexture, samplePos).x;
+    float weight = (1.0 / (0.0001 + abs(depth - pixelDepth)));
+    res.rgb += texture2D(aoMap, vec2(x, y + samplesOffsets[i] * srcTexelSize.y)).rgb * weight;
+    weightSum += weight;
+  }
+  res.rgb /= weightSum;
+
+  #if defined(USE_FOG) && !defined(FOG_TRANSPARENT)
+    // add fog to the result value
+    // goal:    gl_FragColor = fragColor*AO*(1-fogFactor) + fogColor*fogFactor
```
Comment (NataliaDSmirnova) [inline src/gfx/shaders/AOVertBlurWithBlend.frag]: write here: "proper way to get an image with fog and ao requires formula:"

File: src/gfx/shaders/AOVertBlurWithBlend.frag [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    // exists:  AO, fogFactor, fogColor,
```
Comment (NataliaDSmirnova) [inline src/gfx/shaders/AOVertBlurWithBlend.frag]: "But we have already fogged molecule to add AO to. Ok let's split the straight formula into our real steps!"

File: src/gfx/shaders/AOVertBlurWithBlend.frag (hunk 7 above)
Comment (Hadegda) [reply src/gfx/shaders/AOVertBlurWithBlend.frag]: Thanks for your corrections. I modified this comment and really think that now it is more clear)

File: src/gfx/shaders/AOMaterial.js (hunk 6 above)
Comment (Hadegda) [reply src/gfx/shaders/AOMaterial.js]: Thanks, I forgot about it)

File: src/gfx/shaders/AO.frag (hunk 3 above)
Comment (Hadegda) [reply src/gfx/shaders/AO.frag]: Sorry, for confusing you.
There was no specific reason to use in comparison 9th digits after a point. Changed it on usual 0.0000001

File: src/gfx/shaders/AOHorBlur.frag (hunk 4 above)
Comment (Hadegda) [reply src/gfx/shaders/AOHorBlur.frag]: The same answer)

File: src/gfx/shaders/AOHorBlur.frag (hunk 5 above)
Comment (Hadegda) [reply src/gfx/shaders/AOHorBlur.frag]: The same answer)

File: src/gfx/gfxutils.js (hunk 2 above)
Comment (Hadegda) [reply src/gfx/gfxutils.js]: I checked, there is no other place in which we can use it

File: src/gfx/gfxutils.js (hunk 1 above)
Comment (Hadegda) [reply src/gfx/gfxutils.js]: Thanks, I didn't think about it

**Pull request merged**